### PR TITLE
[feat] Add haptic feedback

### PR DESCRIPTION
**Pull request by Frenzie**
References <https://github.com/koreader/koreader/issues/5374>.

<hr>

@pazos This reflects my basic thought process on how to implement this. Most actions wouldn't do anything of course, basically just keyboard & hold. In using [performHapticFeedback](https://developer.android.com/reference/android/view/View.html#performHapticFeedback(int)) it would be up to the system whether it wants to do anything with it.

### CODE REVIEW

File: frontend/apps/reader/modules/readerstatus.lua [hunk 1]
```diff
@@ -1,9 +1,11 @@
 local BookStatusWidget = require("ui/widget/bookstatuswidget")
 local ButtonDialogTitle = require("ui/widget/buttondialogtitle")
+local Device = require("device")
 local InfoMessage = require("ui/widget/infomessage")
 local InputContainer = require("ui/widget/container/inputcontainer")
 local UIManager = require("ui/uimanager")
 local _ = require("gettext")
+local Input = Device.input
```
Comment (Frenzie) [inline frontend/apps/reader/modules/readerstatus.lua]:
```suggestion
```

File: frontend/device/sdl/device.lua [hunk 2]
```diff
@@ -215,11 +215,20 @@ function Device:init()
         setClipboardText = function(text)
             return input.setClipboardText(text)
         end,
+        gameControllerRumble = function(left_intensity, right_intensity, duration)
+            return input.gameControllerRumble(left_intensity, right_intensity, duration)
+        end,
         file_chooser = input.file_chooser,
     }
```
Comment (Frenzie) [inline frontend/device/sdl/device.lua]:
```suggestion
    }

```

File: frontend/device/sdl/device.lua [hunk 3]
```diff
@@ -215,11 +215,21 @@ function Device:init()
         setClipboardText = function(text)
             return input.setClipboardText(text)
         end,
+        gameControllerRumble = function(left_intensity, right_intensity, duration)
+            return input.gameControllerRumble(left_intensity, right_intensity, duration)
+        end,
         file_chooser = input.file_chooser,
-    }
+     }
```
Comment (Frenzie) [inline frontend/device/sdl/device.lua]:
```suggestion
    }
```

File: frontend/apps/reader/modules/readerstatus.lua [hunk 4]
```diff
@@ -38,6 +39,9 @@ function ReaderStatus:addToMainMenu(menu_items)
 end
 
 function ReaderStatus:onEndOfBook()
+    if Device:isHapticFeedbackEnabled() then
```
Comment (Frenzie) [inline frontend/apps/reader/modules/readerstatus.lua]: I think it's probably more elegant if this just reads `Device:performHapticFeedback("CONTEXT_CLICK")`.

Comment (NiLuJe) [reply frontend/apps/reader/modules/readerstatus.lua]: You mean fold the cap check inside performHaptic*?

Comment (Frenzie) [reply frontend/apps/reader/modules/readerstatus.lua]: Exactly. It already defaults as a no-op anyway and this way it just adds a lot of extra boilerplate.

File: frontend/device/sdl/device.lua [hunk 5]
```diff
@@ -215,11 +215,21 @@ function Device:init()
...
         end,
+        gameControllerRumble = function(left_intensity, right_intensity, duration)
+            return input.gameControllerRumble(left_intensity, right_intensity, duration)
+        end,
         file_chooser = input.file_chooser,
     }
 
     self.keyboard_layout = require("device/sdl/keyboard_layout")
 
+    if self.input.gameControllerRumble(0, 0, 0) then
```
Comment (Frenzie) [inline frontend/device/sdl/device.lua]: Note to self: somewhere it should convert -1 into false. Probably best to stick that directly in FFI.

**Comment by Frenzie** [comment]
Merging now. The API can be adjusted if other platforms require it.

**Pull request merged**